src/Microsoft.Health.Fhir.Core/Features/Operations/DataConvert/DataConvertEngine.cs

```diff
@@ -0,0 +1,113 @@
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Core.Configs;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using Microsoft.Health.Fhir.Core.Features.Operations.DataConvert.Models;
+using Microsoft.Health.Fhir.Core.Messages.DataConvert;
+using Microsoft.Health.Fhir.Liquid.Converter;
+using Microsoft.Health.Fhir.Liquid.Converter.Exceptions;
+using Microsoft.Health.Fhir.Liquid.Converter.Hl7v2;
+using Microsoft.Health.Fhir.Liquid.Converter.Models;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.DataConvert
+{
+    public class DataConvertEngine : IDataConvertEngine
+    {
+        private readonly IDataConvertTemplateProvider _dataConvertTemplateProvider;
+        private readonly DataConvertConfiguration _dataConvertConfiguration;
+        private readonly ILogger<DataConvertEngine> _logger;
+
+        private readonly Dictionary<DataConvertInputDataType, IFhirConverter> _dataConverterMap = new Dictionary<DataConvertInputDataType, IFhirConverter>();
+
+        public DataConvertEngine(
+            IDataConvertTemplateProvider dataConvertTemplateProvider,
+            IOptions<DataConvertConfiguration> dataConvertConfiguration,
+            ILogger<DataConvertEngine> logger)
+        {
+            EnsureArg.IsNotNull(dataConvertTemplateProvider, nameof(dataConvertTemplateProvider));
+            EnsureArg.IsNotNull(dataConvertConfiguration, nameof(dataConvertConfiguration));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _dataConvertTemplateProvider = dataConvertTemplateProvider;
+            _dataConvertConfiguration = dataConvertConfiguration.Value;
+            _logger = logger;
+
+            InitializeConvertProcessors();
+        }
+
+        public async Task<DataConvertResponse> Process(DataConvertRequest convertRequest, CancellationToken cancellationToken)
+        {
+            var templateCollection = await _dataConvertTemplateProvider.GetTemplateCollectionAsync(convertRequest, cancellationToken);
+            var result = GetDataConvertResult(convertRequest, new Hl7v2TemplateProvider(templateCollection));
+
+            return new DataConvertResponse(result);
+        }
+
+        private string GetDataConvertResult(DataConvertRequest convertRequest, ITemplateProvider templateProvider)
+        {
+            var dataConverter = _dataConverterMap.GetValueOrDefault(convertRequest.InputDataType);
+            if (dataConverter == null)
+            {
+                // This case should never happen.
+                _logger.LogError("Invalid input data type for conversion.");
+                throw new RequestNotValidException("Invalid input data type for conversion.");
+            }
+
+            try
+            {
+                return dataConverter.Convert(convertRequest.InputData, convertRequest.EntryPointTemplate, templateProvider);
```

Comment:
I notice that this is not passing in a CancellationToken

Reply:
Yes. Currently DotLiquid engine does not accept cancellation token for rendering process. We only configs the "OperationTimeout" setting and it will check whether rendering has timed out.

Reply:
Thanks for raise it.  Actually we are already talking to dotliquid team if it's possible to support cancellation.  I think we can add it into interface first. once underhood component support this feature, we only need to upgrade the nuget package to enable it  without any code changes.